### PR TITLE
ci: Improve accuracy of `wait-for-circleci-workflow-status`

### PR DESCRIPTION
**Pull request by Gudahtt**
## **Description**

The GitHub Actions workflow `wait-for-circleci-workflow-status` has been updated to ensure that it waits for the correct workflow. Previously it always chose the most recent workflow for the given branch, but it may have chosen a workflow corresponding to the wrong commit. It has been updated to find one matching the same commit that triggered the GitHub Actions workflow.

A log has been added to help diagnose any future problems with this workflow, and to help with verification.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/29310?quickstart=1)

## **Related issues**

N/A

## **Manual testing steps**

Check the logs of the "Wait for CircleCI workflow status" job, and see that the workflow ID it's waiting on is correct when making multiple successive commits (comparing the timing using the CircleCI dashboard)

Unfortunately the ID logged in the action is not shown on the CircleCI UI, but you can download the pipeline data with this command: `curl 'https://circleci.com/api/v2/project/gh/MetaMask/metamask-extension/pipeline?branch=[branch]' > pipeline.json` and look through the `pipeline.json` file for an entry that matches the logged ID. The `number` field is the pipeline number, which is in the CircleCI workflow URL (e.g. `https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/[pipeline number]/workflows/[some other number]`)

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by danjm — APPROVED** [review]

File: .github/workflows/wait-for-circleci-workflow-status.yml [hunk 1]
```diff
@@ -13,8 +13,13 @@ jobs:
...
           REPOSITORY: ${{ github.event.repository.name }}
           BRANCH: ${{ github.head_ref || github.ref_name }}
+          # For a `push` event, the HEAD commit hash is `github.sha`.
+          # For a `pull_request` event, `github.sha` is instead the base branch commit hash. The
+          # HEAD commit hash is `pull_request.head.sha`.
+          HEAD_COMMIT_HASH: ${{ github.event.pull_request.head.sha || github.sha }}
         run: |
-          pipeline_id=$(curl --silent "https://circleci.com/api/v2/project/gh/$OWNER/$REPOSITORY/pipeline?branch=$BRANCH" | jq -r ".items[0].id")
+          pipeline_id=$(curl --silent "https://circleci.com/api/v2/project/gh/$OWNER/$REPOSITORY/pipeline?branch=$BRANCH" | jq -r ".items | map(select(.vcs.revision == \"${HEAD_COMMIT_HASH}\" )) | first | .id")
+          echo "Waiting for pipeline '${pipeline_id}' for commit hash '${HEAD_COMMIT_HASH}'"
```
Comment (dbrans) [inline .github/workflows/wait-for-circleci-workflow-status.yml:22]: Would adding the branch to the debug logging help with troubleshooting?

File: .github/workflows/wait-for-circleci-workflow-status.yml [hunk 2]
```diff
@@ -13,8 +13,13 @@ jobs:
           OWNER: ${{ github.repository_owner }}
           REPOSITORY: ${{ github.event.repository.name }}
           BRANCH: ${{ github.head_ref || github.ref_name }}
+          # For a `push` event, the HEAD commit hash is `github.sha`.
+          # For a `pull_request` event, `github.sha` is instead the base branch commit hash. The
+          # HEAD commit hash is `pull_request.head.sha`.
+          HEAD_COMMIT_HASH: ${{ github.event.pull_request.head.sha || github.sha }}
         run: |
-          pipeline_id=$(curl --silent "https://circleci.com/api/v2/project/gh/$OWNER/$REPOSITORY/pipeline?branch=$BRANCH" | jq -r ".items[0].id")
+          pipeline_id=$(curl --silent "https://circleci.com/api/v2/project/gh/$OWNER/$REPOSITORY/pipeline?branch=$BRANCH" | jq -r ".items | map(select(.vcs.revision == \"${HEAD_COMMIT_HASH}\" )) | first | .id")
```
Comment (dbrans) [inline .github/workflows/wait-for-circleci-workflow-status.yml:21]: Just a note - what happens if a CircleCI pipeline was never launched for the specified commit? The current `jq` query would return null in that case. Not sure if this edge case is handled elsewhere in the workflow, but wanted to raise it.

File: .github/workflows/wait-for-circleci-workflow-status.yml (hunk 1 above)
Comment (Gudahtt) [reply .github/workflows/wait-for-circleci-workflow-status.yml:22]: Hmm, potentially. I am not sure of any scenarios where the branch would be ambiguous though.

File: .github/workflows/wait-for-circleci-workflow-status.yml (hunk 2 above)
Comment (Gudahtt) [reply .github/workflows/wait-for-circleci-workflow-status.yml:21]: It's certainly not handled explicitly, but yes it is handled in a sense, both before and after.

If `pipeline_id` is `null`, then the attempt to get `workflow_status` will be unsuccessful, resulting in `workflow_status` being set to `null` as well. The check to compare `workflow_status` to the string "running" will trigger, and the workflow will fail saying that the `workflow_status` is null.

Not ideal, but not terrible as a temporary measure (this workflow is just to help us transition from CircleCI to GitHub Actions).

**Review by dbrans — APPROVED** [review]

**Pull request merged**